cpp/src/nvshmem_plugins/nvshmem_wholememory_bootstrap.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cstddef>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include "nvshmem.h"
+#include "nvshmemi_bootstrap.h"
+#include "nvshmemx.h"
+#include "wholememory/communicator.hpp"
+#include "wholememory/initialize.hpp"
+#include "wholememory/memory_handle.hpp"
+
+#include <wholememory/wholememory.h>
+
+// extern bool nvshmemi_is_nvshmem_bootstrapped;
+
+static wholememory_comm_t bootstrap_comm = nullptr;
+
+static int bootstrap_wholememory_barrier(struct bootstrap_handle* handle)
+{
+  bootstrap_comm->barrier();
+  return WHOLEMEMORY_SUCCESS;
+}
+
+static int bootstrap_wholememory_allgather(const void* sendbuf,
+                                           void* recvbuf,
+                                           int length,
+                                           struct bootstrap_handle* handle)
+
+{
+  bootstrap_comm->host_allgather(sendbuf, recvbuf, length, WHOLEMEMORY_DT_INT8);
+  return WHOLEMEMORY_SUCCESS;
+}
+
+static int bootstrap_wholememory_alltoall(const void* sendbuf,
+                                          void* recvbuf,
+                                          int length,
+                                          struct bootstrap_handle* handle)
+{
+  bootstrap_comm->host_alltoall(sendbuf, recvbuf, length, WHOLEMEMORY_DT_INT8);
+  return WHOLEMEMORY_SUCCESS;
+}
+
+static void bootstrap_wholememory_global_exit(int status)
+{
+  wholememory_error_code_t rc = WHOLEMEMORY_SUCCESS;
+  // TODO:
+  abort();
```

Comment:
Any reason to use `abort` (is this `std::abort`?) over `std::exit`? `std::abort` implies forced exit due to some errors. If this function is to force termination, better rename this function to `..._global_abort()`. Otherwise, why not call std::exit with the passed `status`?

Reply:
It is a boostrap plugin for nvshmem, in the new commit, we refer to the way the mpi boostrap plugin in nvshmem, which use `comm_abort()` function of the comm.